File: app/src/main/java/com/surveysyncengine/domain/api/SurveyApiService.kt
package com.surveysyncengine.domain.api

import com.surveysyncengine.domain.model.SurveyResponse

// ---------------------------------------------------------------------------
// SurveyApiService — domain-level interface for uploading responses.
//
// Lives in domain so SyncEngine depends only on domain, never on data.
// ---------------------------------------------------------------------------

interface SurveyApiService {

    /**
     * Upload a single survey response.
     * Throws on any failure — callers map Throwable → SyncError.
     */
    suspend fun uploadSurveyResponse(response: SurveyResponse): UploadResponseDto

    /**
     * Upload a single media attachment.
     * @return publicly accessible URL where the server stored the file.
     */
    suspend fun uploadAttachment(
        surveyResponseId: String,
        attachmentId: String,
        localFilePath: String,
        mimeType: String,
    ): AttachmentUploadDto
}

data class UploadResponseDto(
    val serverId: String,
    val receivedAt: Long,
)

data class AttachmentUploadDto(
    val attachmentId: String,
    val serverUrl: String,
)

File: ios/SurveySyncEngineIOS/Data/Remote/API/SurveyApiService.swift
//
//  SurveyApiService.swift
//  SurveySyncEngineIOS
//

import Foundation

/// A protocol defining the required network operations for the Agricultural Survey system.
/// This abstraction allows the Sync Engine to remain agnostic of the underlying networking library (e.g., URLSession or Alamofire).
public protocol SurveyApiService {
    
    /// Sends a completed survey response to the remote server.
    /// - Parameter response: The Domain-level model containing all survey answers and metadata.
    /// - Returns: An UploadResponseDto containing the server-assigned ID and confirmation timestamp.
    /// - Throws: An error if the network is unreachable, the request times out, or the server returns a non-200 status code.
    func uploadSurveyResponse(response: SurveyResponse) async throws -> UploadResponseDto
    
    /// Uploads a single media file (image, document, etc.) associated with a specific survey.
    /// - Parameters:
    ///   - surveyResponseId: The unique identifier of the parent survey response.
    ///   - attachmentId: The unique identifier for this specific attachment.
    ///   - localFilePath: The disk location of the file to be uploaded.
    ///   - mimeType: The format of the file (e.g., "image/png") to set the correct HTTP Content-Type.
    /// - Returns: An AttachmentUploadDto containing the final remote URL of the uploaded file.
    /// - Throws: An error if the file cannot be read or the upload fails.
    func uploadAttachment(
        surveyResponseId: String,
        attachmentId: String,
        localFilePath: String,
        mimeType: String
    ) async throws -> AttachmentUploadDto
}
